### ISSUE TITLE
Error: one or more Python packages failed to install [error code 1]

### ISSUE THREAD

**Issue by vfincke** [RockefellerUniversity/Herper#24]
Hi everyone, 

I just started working with Harper in oder to install masc2 in R. 
I set up everything but run into the error that python packages fail to install. Can you help me with this issue? 

Thanks! Best wishes, 
Victoria 

> install_CondaTools(tools="macs2", env="PeakCalling_analysis", 
+                    pathToMiniConda="/Users/victoriafincke/Library/r-miniconda-arm64/", updateEnv = TRUE)
* Using Miniconda at: /Users/victoriafincke/Library/r-miniconda-arm64/
Error: one or more Python packages failed to install [error code 1]

**Comment by vfincke** [comment]
Hi Matt, 

Thanks for the quick response and your offer to help me! 

I tried to install it from GitHub, but that gives me the following error: 
remotes::install_github("RockefellerUniversity/Herper")
library(Herper)
Warning messages:
1: In get(results[[i]], packages[[i]]) :
  internal error -3 in R_decompress1
2: In get(results[[i]], packages[[i]]) :
  internal error -3 in R_decompress1

Installing it via Bioconductor works, but results in the following error when using install_CondaTools: 
> library(Herper)
> install_CondaTools(tools="macs2", env="PeakCalling_analysis",
+                    pathToMiniConda="~/Desktop/fresh_conda")
* No Miniconda found at: /Users/victoriafincke/Desktop/fresh_conda
* Installing Miniconda along with core packages -- please wait a moment ...
https://github.com/conda-forge/miniforge/releases/latest/downloadError in if (file.exists(installer)) return(installer) : 
  argument is of length zero

The link seems to be the problem here, it also doesn’t work when I copy it into the browser. This is also why I tried to install conda separately... 

Thanks again and best wishes,

Victoria 



> Am 21.04.2023 um 20:38 schrieb Matt Paul ***@***.***>:
> 
> install_CondaTools(tools="macs2", env="PeakCalling_analysis",
> pathToMiniConda="~/Desktop/fresh_conda")

**Comment by matthew-paul-2006** [comment]
Hi Victoria,

So after looking around a little bit, I think there are two issues. For the conda install there seems like the conda-forge link is not working appropriately for arm64. I am going to try and get that fixed up. 

For the MACS2 install I think the issue lies with MACS2. It does not seem like there is an appropriate version of MACS2 on conda for your machine. As you can see below, they do have a build for Apple machines, but it is not for the new arm64 containing laptops. If there was it would be an additional entry labeled *osx-arm64*. This will likely be an ongoing issue as bioinformatic tools often do not have as much support as other data analytic tools from other fields (and tools on newer versions will likely just be left behind). 

<img width="552" alt="Screen Shot 2023-05-02 at 6 05 13 PM" src="https://user-images.githubusercontent.com/23037538/235796089-37d4ecab-fe3d-4d2f-9818-3bad24b720e3.png">

As an alternative, you can try running *pip* for the installation like here:https://pypi.org/project/MACS2/. We do this for tools that do not have an appropriate version on Conda. And we can actually still use Herper so that it is all still contained in the Conda architecture. We simply use the function *with_CondaEnv*, so it looks something like this:

```
with_CondaEnv("PeakCalling_analysis",
 system2(command="pip",
  args = c("install", "MACS2"),
  stdout = TRUE),
 pathToMiniConda=myMiniconda)

```
Once it is installed, you should just be able to call MACS2 using a similar command, as if it were installed directly by conda.

Let me know if that works.

**Comment by kmatlik** [comment]
Hi Matt,

I am getting similar issues here.  

When I first tried to install miniconda through Herper, I got the following error:
```
install_CondaTools(tools = "python", env = "test", updateEnv = T)
* No Miniconda found at: C:/Users/kvare/AppData/Local/r-miniconda
* Installing Miniconda along with core packages -- please wait a moment ...
Error: miniconda installation failed [exit code 2]
```


I was able to install miniconda manually and then the following worked:
```
myMiniconda <- file.path("/Users/kvare/miniconda3")
install_CondaTools(tools = "python", env = "test", pathToMiniConda = myMiniconda, updateEnv = T)
```

However, when now trying to install python tools I get the same error as Victoria above:
```
install_CondaTools(tools = "star", env = "test", pathToMiniConda = myMiniconda, updateEnv = T)
Error: one or more Python packages failed to install [error code 1]
```

I tried different tools here (kallisto, salmon, macs2) and wasn't able to install any of them with install_CondaTools(). However, pip, for instance, was installed properly and your code above for installing salmon and star using pip also worked for me. 

So basically I should be able to use these tools now with this workaround, but I'm posting this so you are aware that these errors also came up on my Windows x64. 

I was using Herper_1.10.0 with R version 4.3.1 (2023-06-16 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)

Best,
Kärt

**Comment by matthew-paul-2006** [comment]
Thanks for the extra examples Kärt. I think this is all down to the availability of these packages. Herper is limited by what packages are present in the Conda repository of packages, and Windows systems are often less catered for. All the packages you struggled with do not have builds for Windows available on Conda unfortunately. Fortunately, there are Windows builds elsewhere to fill in these gaps. I'm glad you got Pip to work as the workaround to grab these tools. 

By checking the [Anaconda repository list](https://anaconda.org/anaconda/repo) you can always screen the tools you want to use. Or you can use the conda_search() function to check the tools you want to use.
